### PR TITLE
ddl: unify merging unique and non-unique index for multi-schema change

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/53677

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: tests/integrationtest/t/executor/infoschema_reader.test [hunk 1]
```diff
@@ -130,6 +130,8 @@ connection default;
 create table tt (a int);
 alter table tt add index t(a), add column b int;
 select db_name, table_name, job_type from information_schema.DDL_JOBS limit 3;
+alter table tt add index t2(a), add unique index tb(b), add column c int;
+select db_name, table_name, job_type from information_schema.DDL_JOBS limit 4;
```
Comment (D3Hunter) [inline tests/integrationtest/t/executor/infoschema_reader.test:134]: why not limit 7 to show items for later DDL, else i am not sure what's this part is checking

File: pkg/ddl/reorg.go [hunk 2]
```diff
@@ -212,8 +212,12 @@ func (rc *reorgCtx) getRowCount() int64 {
 // the additional ddl round.
 //
 // After that, we can make sure that the worker goroutine is correctly shut down.
-func (w *worker) runReorgJob(reorgInfo *reorgInfo, tblInfo *model.TableInfo,
-	lease time.Duration, f func() error) error {
+func (w *worker) runReorgJob(
+	reorgInfo *reorgInfo,
+	tblInfo *model.TableInfo,
+	lease time.Duration,
+	reorgJob func() error,
```
Comment (D3Hunter) [inline pkg/ddl/reorg.go:219]:
```suggestion
	reorgFn func() error,
```
reorgJob sounds like a `job`

**Review by D3Hunter — COMMENTED** [review]
will review later

File: tests/integrationtest/t/executor/infoschema_reader.test (hunk 1 above)
Comment (lance6716) [reply tests/integrationtest/t/executor/infoschema_reader.test:134]: I want to check there's only one add index sub-job, so it's really merged. `limit 4` can show the "multi-schema" line of last DDL so we know all lines belongs to this DDL is printed

Comment (D3Hunter) [reply tests/integrationtest/t/executor/infoschema_reader.test:134]: there is a `alter table tt add index t(a), add column b int;` above, the result is the similar😓

Comment (lance6716) [reply tests/integrationtest/t/executor/infoschema_reader.test:134]: this PR tests about merging unique and non-unique add index. I can delete that case in next commit

File: pkg/lightning/backend/local/duplicate.go [hunk 3]
```diff
@@ -1085,7 +1091,13 @@ func (local *DupeController) CollectLocalDuplicateRows(ctx context.Context, tbl
 // CollectRemoteDuplicateRows collect duplicate keys from remote TiKV storage. This keys may be duplicate with
```
Comment (tangenta) [inline pkg/lightning/backend/local/duplicate.go:1091]:
```suggestion
// collectRemoteDuplicateRows collect duplicate keys from remote TiKV storage. This keys may be duplicate with
```

**Review by tangenta — APPROVED** [review]

File: tests/integrationtest/t/executor/infoschema_reader.test (hunk 1 above)
Comment (lance6716) [reply tests/integrationtest/t/executor/infoschema_reader.test:132]: https://github.com/pingcap/tidb/pull/53632/commits/d46188912fbf94940e33ccfb165090122d30ba7c

File: pkg/ddl/backfilling_scheduler.go [hunk 4]
```diff
@@ -422,7 +422,15 @@ func (b *ingestBackfillScheduler) setupWorkers() error {
 	for _, e := range b.reorgInfo.elements {
 		indexIDs = append(indexIDs, e.ID)
 	}
-	engines, err := b.backendCtx.Register(indexIDs, job.TableName)
+	var uniques []bool
+	switch v := job.Args[0].(type) {
+	case *bool:
```
Comment (tangenta) [inline pkg/ddl/backfilling_scheduler.go:427]: `*bool` or `bool`?

Comment (lance6716) [reply pkg/ddl/backfilling_scheduler.go:427]: it's really `*bool` I don't know why 😂

Comment (lance6716) [reply pkg/ddl/backfilling_scheduler.go:427]: very hard to know the exact type of different DDL jobs, hope we can improve it in DDL refactor soon.

Comment (tangenta) [reply pkg/ddl/backfilling_scheduler.go:427]: `TestGlobalSortMultiSchemaChange` failed, that's why I have this question..

```
[2024/05/31 03:39:12.035 +00:00] [ERROR] [misc.go:114] ["panic in the recoverable goroutine"] [label=ddl] [funcInfo=onCreateIndex] [r="runtime error: index out of range [0] with length 0"] [stack="github.com/pingcap/tidb/pkg/util.Recover\n\tpkg/util/misc.go:118\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.goPanicIndex\n\tGOROOT/src/runtime/panic.go:114\ngithub.com/pingcap/tidb/pkg/ddl/ingest.(*litBackendCtx).Register\n\tpkg/ddl/ingest/engine_mgr.go:80\ngithub.com/pingcap/tidb/pkg/ddl.(*ingestBackfillScheduler).setupWorkers\n\tpkg/ddl/backfilling_scheduler.go:433\ngithub.com/pingcap/tidb/pkg/ddl.(*ddlCtx).writePhysicalTableRecord\n\tpkg/ddl/backfilling.go:602\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).addPhysicalTableIndex\n\tpkg/ddl/index.go:1950\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).addTableIndex\n\tpkg/ddl/index.go:1997\ngithub.com/pingcap/tidb/pkg/ddl.runReorgJobAndHandleErr.func1\n\tpkg/ddl/index.go:1056\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).runReorgJob.func1\n\tpkg/ddl/reorg.go:254"] [stack="github.com/pingcap/tidb/pkg/util.Recover\n\tpkg/util/misc.go:114\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.goPanicIndex\n\tGOROOT/src/runtime/panic.go:114\ngithub.com/pingcap/tidb/pkg/ddl/ingest.(*litBackendCtx).Register\n\tpkg/ddl/ingest/engine_mgr.go:80\ngithub.com/pingcap/tidb/pkg/ddl.(*ingestBackfillScheduler).setupWorkers\n\tpkg/ddl/backfilling_scheduler.go:433\ngithub.com/pingcap/tidb/pkg/ddl.(*ddlCtx).writePhysicalTableRecord\n\tpkg/ddl/backfilling.go:602\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).addPhysicalTableIndex\n\tpkg/ddl/index.go:1950\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).addTableIndex\n\tpkg/ddl/index.go:1997\ngithub.com/pingcap/tidb/pkg/ddl.runReorgJobAndHandleErr.func1\n\tpkg/ddl/index.go:1056\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).runReorgJob.func1\n\tpkg/ddl/reorg.go:254"]
```

Comment (lance6716) [reply pkg/ddl/backfilling_scheduler.go:427]: I'll add another case. `*bool` type is also chosen from a test case.

Comment (lance6716) [reply pkg/ddl/backfilling_scheduler.go:430]: https://github.com/pingcap/tidb/pull/53632/commits/f94911d1e05e33aeab1a5234ec1d34fd23f187d7

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/ddl/backfilling_scheduler.go [hunk 5]
```diff
@@ -422,7 +422,22 @@ func (b *ingestBackfillScheduler) setupWorkers() error {
 	for _, e := range b.reorgInfo.elements {
 		indexIDs = append(indexIDs, e.ID)
 	}
-	engines, err := b.backendCtx.Register(indexIDs, job.TableName)
+	var uniques []bool
+	switch v := job.Args[0].(type) {
```
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/ddl/backfilling_scheduler.go:426]: `*[]bool` was caught by `TestGlobalSortMultiSchemaChange`, but I don't know how did it happen.

[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/ddl/backfilling_scheduler.go:426]: https://github.com/pingcap/tidb/pull/53632/commits/0da5755c6ac7b2683e72f8f9e43a5e3da7de193a

We can postpone the debug of understanding `*bool | *[]bool` when refactor `job.Args`

[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/ddl/backfilling_scheduler.go:426]: it's from `DecodeArgs`, we are passing pointer to it, refactor or add a todo both lgtm

**Review by D3Hunter — APPROVED** [review]

**Comment by lance6716** [comment]
/hold

wait to address comment

File: pkg/ddl/backfilling_scheduler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/ddl/backfilling_scheduler.go:426]: json.Unmarshal must require the argument as a pointer (or interface), I think the problem is caused by double-level pointer to bool/[]bool, but I didn't see it for now.

**Comment by lance6716** [comment]
/hold

CI failed by #53789 #53791 #53792

**Pull request merged**